### PR TITLE
A KHC Interpreter???

### PR DESCRIPTION
**Pull request by reilem**
🤔

### CODE REVIEW

File: Backend/FcEvaluate.hs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+-- | Fully evaluates all branches, by calling fullStep
+fullEval :: FcTerm 'Fc -> EvM (Either String (FcTerm 'Fc))
+fullEval t = fullStep t >>= \case
+  Right (FcTmApp t1 t2) -> fullEval t2 >>= \case
+    Right t2' -> fullStep (FcTmApp t1 t2')
+    Left err  -> return $ Left err
+  result -> return $ result
+
+fcEvaluate :: UniqueSupply -> FcProgram 'Fc -> (Either String (FcTerm 'Fc), UniqueSupply)
```
Comment (gkaracha) [inline Backend/FcEvaluate.hs:162]: Best if you describe in the comments here our strategy (why use `collapseProgram`, why use `groundTerm`, etc).

File: Backend/FcEvaluate.hs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+          pure $ Just $ substFcTmInTm sub rhs
+  | otherwise = matchTheAlts dc args rest
+
+
+-- collapseProgram :: FcProgram 'Fc -> FcTerm 'Fc
+-- smallStep :: MonadUnique m => FcTerm 'Fc -> m (Maybe (FcTerm 'Fc))
+
+groundTerm :: FcTerm 'Fc -> FcTerm 'Fc
+groundTerm (FcTmTyAbs ty t1) = FcTmTyApp (FcTmTyAbs ty (groundTerm t1)) FcTyUnit
+groundTerm t                 = t
```
Comment (gkaracha) [inline Backend/FcEvaluate.hs]: I think this one also deserves a comment about how it works. Also, the comment should mention that this is only patched grounding. For example
```haskell
let f = <expr1> in /\a. <expr2>
```
will not be grounded, but this one
```haskell
/\a. let f = <expr1> in <expr2>
```
will (even though both of them could initially have the exact same type). The reason that this problem does not arise in our case is because WE KNOW that our type inference and elaboration algorithm produces type abstraction only at the type level. This is the reason that `groundTerm` suffices in our case, but this should still be explained in a comment.

File: Backend/FcEvaluate.hs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    FcTmERROR err _ -> return $ Left err
+    res             -> return $ Right res
+
+-- | Fully evaluates all branches, by calling fullStep
+fullEval :: FcTerm 'Fc -> EvM (Either String (FcTerm 'Fc))
+fullEval t = fullStep t >>= \case
+  Right (FcTmApp t1 t2) -> fullEval t2 >>= \case
+    Right t2' -> fullStep (FcTmApp t1 t2')
+    Left err  -> return $ Left err
+  result -> return $ result
```
Comment (gkaracha) [inline Backend/FcEvaluate.hs]: You can leave this function as-is for now, but I think that you should explain somewhere about our evaluation strategy (at least a comment).

File: Backend/FcEvaluate.hs [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+      | length args /= length xs -> error "<impossible: elab failed!>"
+      | otherwise -> do
+          sub <- buildFcTmSubst . zipExact xs <$> mapM freshenLclBndrs args
+          pure $ Just $ substFcTmInTm sub rhs
+  | otherwise = matchTheAlts dc args rest
+
+
+-- collapseProgram :: FcProgram 'Fc -> FcTerm 'Fc
+-- smallStep :: MonadUnique m => FcTerm 'Fc -> m (Maybe (FcTerm 'Fc))
+
```
Comment (gkaracha) [inline Backend/FcEvaluate.hs]:
```suggestion
```

**Review by gkaracha — CHANGES_REQUESTED** [review]
This patch might fix a few things, but it doesn't do "the-right-thing". That is OK for now, but please add in comments about all the assumptions that we have made and our overall reasoning for taking this approach. We're gonna be scratching our heads later if we don't do it right away!

TL;DR: Add comments and we're good to go :slightly_smiling_face:

File: Backend/FcEvaluate.hs [hunk 5]
```diff
@@ -112,30 +109,56 @@ matchTheAlts  dc  args ((FcAltFc (FcConPat dc' xs) rhs):rest)
...
-groundTerm :: FcTerm 'Fc -> FcTerm 'Fc
-groundTerm (FcTmTyAbs ty t1) = FcTmTyApp (FcTmTyAbs ty (groundTerm t1)) FcTyUnit
-groundTerm t                 = t
-
--- | Follows lazy operations semantics
+-- | Perform a multi-step evaluation using lazy operational semantics. This
+-- | function performs single-step evaluations until the result of this single
+-- | step evaluation indicates that the given term is fully evaluated. If the
+-- | term is an error, then the error's string is returned, otherwise the
+-- | full term is returned.
```
Comment (gkaracha) [inline Backend/FcEvaluate.hs]:
```suggestion
-- | Perform multi-step evaluation using call-by-name operational semantics. This
-- function repeatedly performs single-step evaluation, until the result is either in
-- Weak Head Normal Form (WHNF), or is an 'FcTmERROR'.
```

File: Backend/FcEvaluate.hs [hunk 6]
```diff
@@ -96,11 +71,33 @@ smallStep (FcTmCaseFc scr alts) = smallStep scr >>= \case
   Just scr' -> pure $ Just $ FcTmCaseFc scr' alts
   Nothing   -> pure $ Nothing
 
+-- | Check whether a term is a user defined data constructor application.
```
Comment (gkaracha) [inline Backend/FcEvaluate.hs]:
```suggestion
-- | Check whether a term is a user-defined data constructor application.
```

File: Backend/FcEvaluate.hs [hunk 7]
```diff
@@ -96,11 +71,33 @@ smallStep (FcTmCaseFc scr alts) = smallStep scr >>= \case
...
+isUserDefinedDataConApp :: FcTerm 'Fc -> Bool
+isUserDefinedDataConApp = isJust . userDefinedDataConAppMaybe
+{-# INLINE isUserDefinedDataConApp #-}
+
+-- | Check whether a term is a user defined data constructor application
+-- | and return its parts.
```
Comment (gkaracha) [inline Backend/FcEvaluate.hs]:
```suggestion
-- | Check whether a term is a user defined data constructor application
-- and return its parts.
```
General comment: starting the first line of a comment block with `-- |` is sufficient for getting haddock-compatible documentation. You don't have to add a bar on every line of the comment. :wink:

File: Backend/FcEvaluate.hs [hunk 8]
```diff
@@ -96,11 +71,33 @@ smallStep (FcTmCaseFc scr alts) = smallStep scr >>= \case
...
+
+-- | TODO: remove
 matchTheUnit :: [FcAlt 'Fc] -> EvM (Maybe (FcTerm 'Fc))
 matchTheUnit []                              = error "<unit impossible: pm compiler failed!>"
 matchTheUnit ((FcAltFc FcUnitPat     rhs):_) = pure $ Just rhs
 matchTheUnit ((FcAltFc (FcConPat {}) _  ):_) = error "<unit impossible: elab failed!>"
 
+-- | Matches the given data constructor to one of the given alternatives, and
+-- | substitutes the variables in the alternative's right hand side with the
+-- | data constructor's arguements.
```
Comment (gkaracha) [inline Backend/FcEvaluate.hs]:
```suggestion
-- | Matches the given data constructor to one of the given alternatives, and
-- substitutes the variables in the alternative's right-hand side with the
-- data constructor's arguments.
```

File: Backend/FcEvaluate.hs [hunk 9]
```diff
@@ -142,25 +144,24 @@ groundTerm :: FcTerm 'Fc -> FcTerm 'Fc
...
 groundTerm t                 = t
 
--- | Fully evaluates given term using lazy semantics. However if the final
--- | result is a term application, then the right term in the application may
--- | not be fully evaluated. In order to produce a clear, readable output we
--- | choose to evaluate all branches recursively before returning the result.
+-- | Fully evaluates given term. If final result is a term application, then
+-- | the right hand side term in the application may not be fully evaluated.
+-- | In order to produce a clear, readable output we choose to evaluate all
+-- | branches recursively before returning the result.
```
Comment (gkaracha) [inline Backend/FcEvaluate.hs]:
```suggestion
-- | Fully evaluates a given term.
```

File: Backend/FcEvaluate.hs [hunk 10]
```diff
@@ -142,25 +144,24 @@ groundTerm :: FcTerm 'Fc -> FcTerm 'Fc
...
 fullEval :: FcTerm 'Fc -> EvM (Either String (FcTerm 'Fc))
 fullEval t = fullStep t >>= \case
   Right (FcTmApp t1 t2) -> fullEval t2 >>= \case
     Right t2' -> fullStep (FcTmApp t1 t2')
     Left err  -> return $ Left err
   result -> return $ result
```
Comment (gkaracha) [inline Backend/FcEvaluate.hs]: Actually, I think this function is wrong. A data constructor can have multiple arguments. We should start evaluating them from left to right. You should try and recursively evaluate `t1` first, and then `t2`.

**Review by gkaracha — CHANGES_REQUESTED** [review]

File: Backend/FcEvaluate.hs [hunk 11]
```diff
@@ -112,30 +109,56 @@ matchTheAlts  dc  args ((FcAltFc (FcConPat dc' xs) rhs):rest)
...
+-- | Grounds the given term by wrapping all top level type abstractions in a
+-- | type application on unit. This is only patched grounding, and so no
+-- | internal type abstractions are grounded. For example:
+-- |   /\ a. let f = <expr1> in <expr2>
+-- | is grounded to:
+-- |   (/\ a. let f = <expr1> in <expr2>) Unit
+-- |
+-- | But (let f = <expr1> in /\a. <expr2>) is left unchanged. We do this
+-- | because we know the elaboration & desugaring phases only produce top level
+-- | type abstractions. So this method will be sufficient in all cases.
```
Comment (gkaracha) [inline Backend/FcEvaluate.hs]:
```suggestion
-- | Grounds the given term by looking for top-level type abstractions, and
-- then applying it on as many unit values as needed. This is only patched
-- grounding, and so no internal type abstractions are grounded. For example:
--
-- > /\ a. let f = <expr1> in <expr2>
--
-- is grounded to:
--
-- > (/\ a. let f = <expr1> in <expr2>) Unit
-- 
-- But @(let f = <expr1> in /\a. <expr2>)@ is left unchanged. We do this
-- because we know the elaboration and desugaring phases only produce top-level
-- type abstractions. So this method will be sufficient in all cases.
-- TODO: In the future, this function should be type-directed: the type should
-- determine whether we should create an application to unit or not.
```
@reilem For haddock-compatible markup comments look here: https://www.haskell.org/haddock/doc/html/ch03s08.html

**Review by gkaracha — CHANGES_REQUESTED** [review]

File: Backend/FcEvaluate.hs (hunk 10 above)
Comment (reilem) [reply Backend/FcEvaluate.hs]: I don't think that is necessary because the congruence rule will evaluate `t1` till it goes no further. Once `t1` is in normal form we receive the `FcTmApp` here.

Comment (reilem) [reply Backend/FcEvaluate.hs]: Added a note
